Add details about non-migratable VMs (#849)
harvester/docs: docs/vm/live-migration.md

  <link rel="canonical" href="https://docs.harvesterhci.io/v1.5/vm/live-migration"/>
</head>

Live migration means moving a virtual machine to a different host without downtime. A couple of comprehensive processes and tasks are done under the hood to fulfill the live migration.

> The general requirements are:

Reviewer: ```suggestion
The general requirements for suitable migration target nodes are:
```

- The cluster has at least one additional scheduable node which can meet all the scheduling rules of the VM besides the current host node.

- The destination node has enough spare resources to host the VM.

- The requested CPU, memory, [volumes](./create-vm.md#volumes), devices and other resources of the VM can be copied or rebuilt on the destination host while the source VM is still running.

## Non-migratable VMs

The definitions of VM are versatile, a VM cannot perform live migration when one or more of following conditions are met.

Remove the related device or add more schedulable nodes can make the VM live-migratable.

### Has non-migratable devices or node-selector

- The VM has any volume of the `CD-ROM` type.

- The VM has any volume of the `Container Disk` type.

- The VM has any volume with `volumeAccessMode` `ReadWriteOnce`.

- The VM has any volume and it's parent `StorageClass` has `replica 1`.

  ::caution
  This condition is not detected on all cases.
  ::

> - The VM has `PCI passthrough` or `vGPU` devices.

Reviewer 2: ```suggestion
- The VM has host devices passthrough, such as `PCI` or `vGPU`.
```

- The VM has a [node selector](./create-vm.md#node-scheduling) that binds it to a specific node.

### Has scheduling rules which can only match one node

Following conditions are checked on the runtime (e.g. before an upgrade) to mark the VM is non-migratable if only one node matches.

- The VM is on a `cluster network` which spreads to only one node.

  See [Automatically Applied Affinity Rules](./create-vm.md#related-networking-concepts) for more details.

- The VM has `cpu-pinning` enabled and there is only one node enables `CPU Manager`.

  See [Automatically Applied Affinity Rules](./create-vm.md#related-cpu-pinning-concepts) for more details.

- Other [node scheduling](./create-vm.md#node-scheduling) rules.

## Live-migratable VMs

> Besides `non-migratable VMs`, the rest of the running VMs are considered `live-migratable`.

Reviewer: I find the info in this section makes sense only if I read this page in a top-down manner. Landing on this section via a href makes the info feels incomplete. Most other documents seems to reference the non-migratable VMs section because understandably, they are the ones that impact the user. Is it necessary for other documents to reference a "live-migratable VMs" section since there is nothing interesting about these VMs. Instead of sections for non-migratable and live-migratable VMs, can we use a more generic title that says something like "Criteria For Live Migration"?

A virtual machine is considered non-migratable if it has one or more of the following:

- Volume with the following properties:

    - Type: `CD-ROM` or `Container Disk`
    - Access mode: `ReadWriteOnce`
    - StorageClass replica count: `1` (This is not detected in all cases.)

- Host devices passthrough such as `PCI` and `vGPU`

- [Node selector](./create-vm.md#node-scheduling) that binds the virtual machine to a specific node

- [Scheduling rules](./create-vm.md#node-scheduling) that match only one node

  The following are examples of rule conditions that are checked at runtime. For more information, see [Automatically Applied Affinity Rules](./create-vm.md#related-networking-concepts).

    - The virtual machine is attached to a cluster network that covers only one node.

    - CPU pinning is enabled on the virtual machine, and CPU Manager is only enabled on one node.

:::note

To live-migrate the virtual machine, you must first remove non-migratable devices and add schedulable nodes.

:::

## Live-migratable Virtual Machines

Besides the `non-migratable virtual machines`, the rest of the running virtual machines are considered `live-migratable`.

## How Migration Works

> ### The `VirtualMachineInstanceMigration` Object

Reviewer: I feel like this is getting into the weed. What should the reader do with this info?

Can we just add a link to reference the comprehensive KB on live migration at https://harvesterhci.io/kb/vm_live_migration_policy_and_configuration? Feels like the KB might be a better home for these  details.

Author: The KB was written around two years ago, and now many changes have been introduced:

Here the `second-level` title is `How Migration Works`, and it just mentions `cpu model matching`, adding the `the vmim object` expands it.

There are 4 ways to trigger migration on v1.6.0:  **Migrate** menu, node maintenance, upgrade, hot-plug, and each has some characters. 

Both of doc && kb need updates to describe things from different perspective, and cross-refer.
 
 Though some document setcions seem redundant, but it is also true that not every user has the patience to read all.
 
Certain degree of in-depth description of `live migration` on `live-migration.md` does not harm, at least.

When a virtual machine migration action is triggered, a `VirtualMachineInstanceMigration` object is created. The controller works on this object.

The cross-referrences between the `VirtualMachineInstanceMigration` and `VirtualMachineInstance` objects are:

`VirtualMachineInstance` .status.migrationState.migrationUID = `VirtualMachineInstanceMigration`.UID

`VirtualMachineInstanceMigration` .spec.VMIName = `VirtualMachineInstance`.Name

The `VirtualMachineInstanceMigration` object's name varies:

#### When the Migration is Manually Triggered

When a migration is triggered from the **Migrate** [menu item](#starting-a-migration), the format of `VirtualMachineInstanceMigration` object's name is:

- `Virtual machine name` + `-` + `a random string`

  Example: `vm1-a3d1f`

### When the Migration is Automatically Triggered

When a migration is triggered [automatically](#automatically-triggered-batch-migrations), the format of `VirtualMachineInstanceMigration` object's name is:

- `kubevirt-evacuation-` + `a random string`

>   Example: `kubevirt-evacuation-9c485`

Author: Looks good!

## Non-migratable Virtual Machines

A virtual machine is considered non-migratable if it has one or more of the following:

- Volume with the following properties:

    - Type: `CD-ROM` or `Container Disk`
    - Access mode: `ReadWriteOnce`
    - StorageClass replica count: `1` (This is not detected in all cases.)

- Host devices passthrough such as `PCI` and `vGPU`

- [Node selector](./create-vm.md#node-scheduling) that binds the virtual machine to a specific node

- [Scheduling rules](./create-vm.md#node-scheduling) that match only one node

  The following are examples of rule conditions that are checked at runtime. For more information, see [Automatically Applied Affinity Rules](./create-vm.md#related-networking-concepts).

    - The virtual machine is attached to a cluster network that covers only one node.

    - CPU pinning is enabled on the virtual machine, and CPU Manager is only enabled on one node.

:::note

To live-migrate the virtual machine, you must first remove non-migratable devices and add schedulable nodes.

:::

## Live-migratable Virtual Machines

> Besides the `non-migratable virtual machines`, the rest of the running virtual machines are considered `live-migratable`.

Reviewer 3: ```suggestion
Virtual machines that do not have the properties of non-migratable virtual machines can be live-migrated.
```

## How Migration Works

> ### The `VirtualMachineInstanceMigration` Object

Reviewer 3: ```suggestion
### `VirtualMachineInstanceMigration` Object
```

> When a virtual machine migration action is triggered, a `VirtualMachineInstanceMigration` object is created to track the state and progress of the operation. The Harvester controller correlates the `VirtualMachineInstanceMigration` object with the `VirtualMachineInstance` object by ensuring the instance object's identity are reflected in the migration object.
> 
> For example, during the migration of a virtual machine named `demo`, the associated migration object's UID is added to the instance object's `.status.migrationState.migrationUID` property:

Reviewer 3: ```suggestion
When a virtual machine migration action is triggered, a `VirtualMachineInstanceMigration` object is created to track the state and progress of the operation. The Harvester controller correlates the `VirtualMachineInstanceMigration` object with the `VirtualMachineInstance` object by ensuring the instance object's identity is reflected in the migration object.

In the following example, the virtual machine named `demo` has an associated migration object. The UID of this object is added to the instance object's `.status.migrationState.migrationUID` property during migration.
```

```sh

$ kubectl get vmi demo -ojsonpath={.status.migrationState.migrationUID}

1d6d7273-275d-48e0-bb76-62e240b42aaf

```

Conversely, the instance object's name is added to the migration object's `.spec.vmiName` property:

```sh

$ kubectl get vmim demo-6crrk -ojsonpath={.spec.vmiName}

demo

```

> The format of the `VirtualMachineInstanceMigration` object's name varies depending on whether the migration is manually or automatically triggered.
> 
> When a migration is triggered from the **Migrate** [menu item](#starting-a-migration), the `VirtualMachineInstanceMigration` object's name is prefixed with the virtual 
> machine's name followed by a random string. For example, `vm1-a3d1f`.
> 
> When a migration is triggered [automatically](#automatically-triggered-batch-migrations), the `VirtualMachineInstanceMigration` object's name is prefixed with `kubevirt-evacuation-` followed by a random string. For example, `kubevirt-evacuation-9c485`.

Reviewer 3: ```suggestion
The format of the `VirtualMachineInstanceMigration` object's name varies depending on whether the migration is manually or automatically triggered.

When a migration is triggered from the **Migrate** [menu item](#starting-a-migration), the `VirtualMachineInstanceMigration` object's name is prefixed with the virtual 
machine's name and a random string (for example, `vm1-a3d1f`).

When a migration is triggered [automatically](#automatically-triggered-batch-migration), the `VirtualMachineInstanceMigration` object's name is prefixed with `kubevirt-evacuation-` and a random string (for example, `kubevirt-evacuation-9c485`).
```

> :::note
> 
> Harvester UI does not provide direct information about the source of the migration, you need to check the name of `VirtualMachineInstanceMigration` object to know the action source of a migration.
> 
> :::
> 
> ### CPU Models Matching

Reviewer 3: ```suggestion
:::note

The Harvester UI does not specify the source of the migration. You must check the name of `VirtualMachineInstanceMigration` object to retrieve this information.

:::

### CPU Model Matching
```

1. Go to the **Virtual Machines** page.
1. Find the virtual machine in migrating status that you want to abort. Select **⋮ > Abort Migration**.

:::note

- The `Abort Migration` menu is available when the VM already has a running or pending migration process.

- Don't click `Abort Migration` if it is created by the [batch-migrations](#automatically-triggered-batch-migrations)

:::

> ## Automatically triggered batch-migrations

Reviewer: I am not sure if this section is necessary. Is this a general FYI for the readers?

Reviewer: Seems to me that you want to differentiate between system-initiated vs. user-initiated migration, right? Can you elaborate more on why the action source is important to the user and what the user can do with that info? What user decisions will be affected by this knowledge? E.g., is it the case that we don't want user to terminate system-initiated migration?

1. Go to the **Virtual Machines** page.
1. Find the virtual machine in migrating status that you want to abort. Select **⋮ > Abort Migration**.

:::note

> - The **Abort Migration** menu item is available when the virtual machine already has a running or pending migration process.
> 
> - Don't click `Abort Migration` if it is created by the [batch-migrations](#automatically-triggered-batch-migrations). See [The `VirtualMachineInstanceMigration` Object](#the-virtualmachineinstancemigration-object) for more details.

Reviewer 3: ```suggestion
The **Abort Migration** menu item is already running or has a pending migration process.

Do not use this UI feature if the migration process was created using [batch migration](#automatically-triggered-batch-migration). For more information, see [`VirtualMachineInstanceMigration` Object](#virtualmachineinstancemigration-object).
```

Author: > The **Abort Migration** menu item is already running or has a pending migration process.

This description is not accurate, it needs to express below information:

When the VM has already a related virtualmachineinstancemigration object (created by **Migrate** menu or batch-migration or others) and the latter is in state `running` or `pending` (it has other state like `failed`, `successful`, `aborted`), then the VM has the menu **Abort Migration**.

:::

> ## Automatically triggered batch-migrations
> 
> Both [Harvester upgrade](../upgrade/automatic.md#virtual-machine-management-through-the-upgrade) and [node maintenance](../host/host.md#node-maintenance) benefit from the **Live Migration**, and the process is slightly different with above [Starting a Migration](#starting-a-migration). It is called `batch-migrations`.
> 
> The general process is:

Reviewer 3: ```suggestion
## Automatically Triggered Batch Migration

[Harvester upgrades](../upgrade/automatic.md#live-migratable-vms) and [node maintenance](../host/host.md#node-maintenance) both benefit from live migration. The underlying process, which is called *batch migration*, is slightly different from the one described in [Starting a Migration](#starting-a-migration). This process involves the following steps:
```

> 1. The controller watchs a dedicated taint on the node object.
> 
> 1. The controller creates a `VirtualMachineInstanceMigration` object for each [live-migratable virtual machine](#live-migratable-virtual-machines) on the current node.
> 
> 1. The migrations are queued, scheduled internally, and are process in batch mode. UI shows `Pending migration` or `Migrating` according to their status.
> 
> 1. The controller monitors the processing and waits until all of them are done or time-out.

Reviewer 3: ```suggestion
1. The controller watches a dedicated taint on the node object.

1. The controller creates a `VirtualMachineInstanceMigration` object for each [live-migratable virtual machine](#live-migratable-vms) on the current node.

1. The migrations are queued, scheduled internally, and processed in batches. The Harvester UI shows the statuses **Pending migration** and **Migrating** to indicate progress.

1. The controller monitors the processing and waits until all are completed or have timed out.
```

- Cluster level: Run `kubectl edit kubevirts.kubevirt.io -n harvester-system` and add `spec.configuration.cpuModel: "123"`. This change also affects newly created VMs.
- Individual VMs: Modify the VM configuration to include `spec.template.spec.domain.cpu.model: "123"`.

> Both methods require to restart the VMs. If you are certain that all nodes in the cluster support a specific CPU model, you can define this at the cluster level before creating any VMs. In doing so, you eliminate the need to restart the VMs (to assign the CPU model) during live migration.

Reviewer 3: ```suggestion
Both methods require you to restart the VMs. If you are certain that all nodes in the cluster support a specific CPU model, you can define this at the cluster level before creating any VMs. In doing so, you eliminate the need to restart the VMs (to assign the CPU model) during live migration.
```
